r/tests/testthat/test-dplyr.R

```diff
@@ -256,6 +262,29 @@ test_that("filter() with %in%", {
   )
 })
 
+test_that("filter() with string ops", {
+  expect_dplyr_equal(
+    input %>%
+      filter(dbl > 2, toupper(chr) %in% c("D", "F")) %>%
+      collect(),
+    tbl
+  )
```

Comment:
This test would pass without the changes in this PR. Do you want to try to test `toupper()` and `tolower()` in a way that would have failed before but succeeds now?

Reply:
It's tricky with the R string functions like this that start with 

```
    if (!is.character(x)) 
        x <- as.character(x)
```

Maybe we can test that by defining `as.character <- function(...) stop("🙅")` in the test?

Reply:
I looked into this. You can't just define `as.character` in the test because you need to replace the one that `toupper` calls. You can't use `testthat::with_mock` to mock base functions anymore, and my other hack of using `trace` to insert a crash seems to crash too many things (presumably legitimate places where as.character gets called elsewhere in the dplyr code). `mockery` might allow this more targeted but I haven't had great success setting it up before; there are other ways we could instrument our code to make sure the Arrow function is called, but I'm not sure it's worth it. It _definitely_ won't work if you do this on a Dataset, so we get some related test coverage there.

Reply:
I did work out a test for this, PTAL